azurelinuxagent/pa/provision/default.py

```diff
@@ -57,45 +58,71 @@ def run(self):
+                logger.info("Provisioning and signaling already completed, skipping.")
+
+            elif provision_state == "provisioned_not_signaled":
+                logger.info("provisioning complete, not signaled")
+                logger.info("signaling...")
+                if self.report_ready():
+                    self.write_signaled()
+                    logger.info("signaling complete")
+                    self.osutil.eject_dvd()
+                else:
+                    logger.info("signaling incomplete")
 
-            logger.info("Running default provisioning handler")
+            else:
+                if provision_state != "not_provisioned":
+                    logger.warn("unexpected previsioning state {0}", provision_state)
 
-            if not self.validate_cloud_init(is_expected=False):
-                raise ProvisionError("cloud-init appears to be running, "
-                                        "this is not expected, cannot continue")
+                logger.info("Running default provisioning handler")
 
-            logger.info("Copying ovf-env.xml")
-            ovf_env = self.protocol_util.copy_ovf_env()
+                if not self.validate_cloud_init(is_expected=False):
+                    raise ProvisionError("cloud-init appears to be running, "
+                                            "this is not expected, cannot continue")
 
-            self.protocol_util.get_protocol(by_file=True)
-            self.report_not_ready("Provisioning", "Starting")
-            logger.info("Starting provisioning")
+                logger.info("Copying ovf-env.xml")
+                ovf_env = self.protocol_util.copy_ovf_env()
 
-            self.provision(ovf_env)
+                self.protocol_util.get_protocol(by_file=True)
+                self.report_not_ready("Provisioning", "Starting")
+                logger.info("Starting provisioning")
 
-            thumbprint = self.reg_ssh_host_key()
-            self.osutil.restart_ssh_service()
+                self.provision(ovf_env)
 
-            self.write_provisioned()
+                thumbprint = self.reg_ssh_host_key()
+                self.osutil.restart_ssh_service()
 
-            self.report_event("Provisioning succeeded ({0}s)".format(self._get_uptime_seconds()),
-                is_success=True,
-                duration=elapsed_milliseconds(utc_start))
+                self.write_provisioned()
+                logger.info("Provisioning complete")
 
-            self.handle_provision_guest_agent(ovf_env.provision_guest_agent)
 
-            self.report_ready(thumbprint)
-            logger.info("Provisioning complete")
+                self.report_event("Provisioning succeeded ({0}s)".format(self._get_uptime_seconds()),
+                    is_success=True,
+                    duration=elapsed_milliseconds(utc_start))
+
+                self.handle_provision_guest_agent(ovf_env.provision_guest_agent)
```

Comment:
The very last thing that happens during provisioning should be the call to write_provisioned(); a crash immediately after that call would mean that, upon restart, provision_state == "provisioned_not_signaled" and the VM would immediately send the "done provisioning" signal to the control plane. However, as this is written, lines 110-117 are executed after write_provisioned(). Unless handle_provision_guest_agent() doesn't do any provisioning work (which I'm pretty sure is not a true statement), this code violates the expectation that provisioning is DONE when write_provisioned() is called.

Reply:
Turns out that handle_provision_guest_agent() does NOT do any provisioning work, just writes some log stuff. Weird name for a method that just logs.

I withdraw this objection.

---
In reply to: [284935536](https://github.com/Azure/WALinuxAgent/pull/1528#discussion_r284935536) [](ancestors = 284935536)